exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/ConvertCountToDirectScan.java

```diff
@@ -85,109 +91,231 @@ protected ConvertCountToDirectScan(RelOptRuleOperand rule, String id) {
-            index = ((RexInputRef) proj.getProjects().get(index)).getIndex();
-          } else {
-            return;  // do not apply for all other cases.
-          }
-        }
+    final CountsCollector countsCollector = new CountsCollector(settings);
+    // if counts were not collected, rule won't be applied
+    if (!countsCollector.collect(agg, scan, project)) {
+      return;
+    }
 
-        String columnName = scan.getRowType().getFieldNames().get(index).toLowerCase();
+    final RelDataType scanRowType = constructDataType(agg);
 
-        cnt = oldGrpScan.getColumnValueCount(SchemaPath.getSimplePath(columnName));
-        if (cnt == GroupScan.NO_COLUMN_STATS) {
-          // if column stats are not available don't apply this rule
-          return;
-        }
-      } else {
-        return; // do nothing.
-      }
+    final DynamicPojoRecordReader<Long> reader = new DynamicPojoRecordReader<>(
+        buildSchema(scanRowType.getFieldNames()),
+        Collections.singletonList(countsCollector.getCounts()));
 
-      RelDataType scanRowType = getCountDirectScanRowType(agg.getCluster().getTypeFactory());
+    final ScanStats scanStats = new ScanStats(ScanStats.GroupScanProperty.EXACT_ROW_COUNT, 1, 1, scanRowType.getFieldCount());
+    final GroupScan directScan = new MetadataDirectGroupScan(reader, oldGrpScan.getFiles(), scanStats);
 
-      final ScanPrel newScan = ScanPrel.create(scan,
-          scan.getTraitSet().plus(Prel.DRILL_PHYSICAL).plus(DrillDistributionTrait.SINGLETON), getCountDirectScan(cnt),
-          scanRowType);
+    final ScanPrel newScan = ScanPrel.create(scan,
+        scan.getTraitSet().plus(Prel.DRILL_PHYSICAL).plus(DrillDistributionTrait.SINGLETON), directScan,
+        scanRowType);
 
-      List<RexNode> exprs = Lists.newArrayList();
-      exprs.add(RexInputRef.of(0, scanRowType));
+    final ProjectPrel newProject = new ProjectPrel(agg.getCluster(), agg.getTraitSet().plus(Prel.DRILL_PHYSICAL)
+        .plus(DrillDistributionTrait.SINGLETON), newScan, prepareFieldExpressions(scanRowType), agg.getRowType());
 
-      final ProjectPrel newProj = new ProjectPrel(agg.getCluster(), agg.getTraitSet().plus(Prel.DRILL_PHYSICAL)
-          .plus(DrillDistributionTrait.SINGLETON), newScan, exprs, agg.getRowType());
+    call.transformTo(newProject);
+  }
 
-      call.transformTo(newProj);
+  /**
+   * For each aggregate call creates field with "count$" prefix and bigint type.
+   * Constructs record type for created fields.
+   *
+   * @param aggregateRel aggregate relation expression
+   * @return record type
+   */
+  private RelDataType constructDataType(DrillAggregateRel aggregateRel) {
+    List<RelDataTypeField> fields = new ArrayList<>();
+    for (int i = 0; i < aggregateRel.getAggCallList().size(); i++) {
+      RelDataTypeField field = new RelDataTypeFieldImpl("count$" + i, i, aggregateRel.getCluster().getTypeFactory().createSqlType(SqlTypeName.BIGINT));
```

Comment:
The fieldname in RelDataTypeField "count$0", "count$1" is not very informative. I think we could either get the field name from the aggregator operator, or use "count$colname" to indicate what column's count it represents.